tests/functional/base.py

```diff
@@ -0,0 +1,316 @@
+                cli_dir=self.config['cli_dir'],
+                username=auth['username'],
+                password=auth['password'],
+                tenant_name=auth['project_name'],
+                user_domain_name=auth['user_domain_name'],
+                project_domain_name=auth['project_domain_name'],
+                identity_api_version='3',
+                uri=auth['auth_url'])
+        
+    def _init_dummy_cloud(self, name, role):
+        if role not in ('owner', 'lessee'):
+            raise Exception('unknown role: %s' % role)
+
+        cloud = {}
+        for field in 'project', 'user':
+            cloud[field] = { 'name': data_utils.rand_name('esi-%s-%s'
+                    % (field, name)) }
+
+        def initialize(field, params):
+            output = self.clients['admin'].openstack('%s create %s' %
+                    (field, cloud[field]['name']), '', params)
+            cloud[field]['id'] = self.parse_details(self, output)['id']
+
+        pw = data_utils.rand_password()
+        initialize('project', '--domain default --enable')
+        initialize('user', '--domain default --password %s --enable' % pw)
+        self.clients['admin'].openstack('role add', '',
+                '--user %s --project %s esi_leap_%s' % (cloud['user']['name'],
+                    cloud['project']['name'], role))
+
+        self.client_info[name] = {
+                'auth_type': 'password',
+                'auth_url': self.client_info['admin']['auth_url'],
+                'username': cloud['user']['name'],
+                'user_id': cloud['user']['id'],
+                'password': pw,
+                'project_name': cloud['project']['name'],
+                'project_id': cloud['project']['id'],
+                'user_domain_name': 'default',
+                'project_domain_name': 'default'}
+
+        for field in cloud.keys():
+            self._cleanups.append([self.clients['admin'].openstack, 
+                    '%s delete' % field, '', '%s' % cloud[field]['name']])
+        self._cleanups.append([self.clients['admin'].openstack, 'role remove', '',
+                '--user %s --project %s esi_leap_%s' % (cloud['user']['name'],
+                    cloud['project']['name'], role)])
+        return 0
+
+    def _init_roles(self):
+        for role in 'esi_leap_owner', 'esi_leap_lessee':
+            try:
+                self.clients['admin'].openstack('role show', '', '%s' % role)
+            except:
+                self.clients['admin'].openstack('role create', '', '%s' % role)
+                self._cleanups.append([self.clients['admin'].openstack,
+                    'role delete', '', '%s' % role])
+
+    def _new_dummy_node(self, owner):
+        node_dir = self.config['dummy_node_dir']
+        if not os.path.exists(node_dir):
+            os.mkdir(node_dir)
+        if not os.path.isdir(node_dir):
+            raise OSError('Error creating dummy node @ %s: not a directory'
+                    % node_dir)
```

Comment:
well, if `os.mkdir(node_dir)` fails then it would raise the appropriate error, do you really need to explicitly test it and raise an OSError manually?

Reply:
....very true. I think I did it this way because I was trying to translate shell script -> python and failed to clean up a lot of things that don't make sense if you think about it too hard.

Reply:
Oh wait, on *another* read thru this case is here for a situation like this:
- some program makes a file called /tmp/nodes
- tests are run
- since os.path.exists('/tmp/nodes') returns true we don't make the directory

I could still remove this check & I think tempfile.mkstemp() would throw an exception later but this feels more useful.